### PR TITLE
Reduce Linux normalization overhead with UTF-16

### PR DESCRIPTION
**Pull request by JinneeJ**
This PR reduces Linux normalization overhead with UTF-16.

Linux normalization currently converts each Kotlin string to UTF-8, copies the input into native memory, allocates a native result, and converts that result back to a Kotlin string.

Using `u16_normalize` with a pinned Kotlin string and a reusable output buffer reduces normalization overhead while preserving the public API and the results of all four normalization forms.

## Implementation

- Pass the input through `String.usePinned`, avoiding the input copy and UTF-8 conversion.
- Reuse a thread-local `CharArray(4096)` for output, approximately 8 KiB per thread. Larger results use libunistring's own allocation and are freed after copying into a Kotlin string.
- Rely on libunistring replacing unpaired surrogates with U+FFFD, which matches the previous UTF-8 encoding behavior.
- Add Linux tests for malformed UTF-16, output buffer boundaries and expansion, buffer reuse, every BMP code unit, randomized strings, and concurrent calls.

## Performance

Measured on Linux ARM64 with Kotlin/Native 2.4.0 release optimization and libunistring 1.0-2. Both implementations run in the same process, with five warm-ups and 21 interleaved rounds. Each sample processes 10,000 strings with 20,000 normalization calls. Input construction and result comparisons are outside the timed section.

Combined text preparation (`NFD → remove combining marks → NFC → lowercase`), median milliseconds:

| Dataset | Original UTF-8 | Pinned UTF-16 | Time reduction |
|---|---:|---:|---:|
| Short ASCII | 23.04 | 16.65 | 28% |
| Short Cyrillic | 30.95 | 20.81 | 33% |
| Short accented Latin | 37.37 | 29.44 | 21% |
| Short CJK and emoji | 16.90 | 9.61 | 43% |
| Short malformed UTF-16 | 18.78 | 13.54 | 28% |
| Mixed text, 512 UTF-16 units | 543.02 | 417.29 | 23% |

The normalization pair alone takes approximately 51 to 67% less time across these datasets. These are microbenchmarks, not end-to-end application measurements.

## Compatibility

Preserving malformed UTF-16 behavior requires **libunistring 0.9.7 or newer**, released in December 2016. In 0.9.6, surrogate-pair validation depended on `CONFIG_UNICODE_SAFETY`; 0.9.7 made those checks unconditional.

## Validation

Before the differential tests were removed ([ref](https://github.com/Doist/doistx-normalize/pull/147#issuecomment-5734079937)), the final implementation matched the original UTF-8 implementation in 645,164 comparisons on Linux ARM64 with libunistring 1.0-2, and in another 645,164 with `latin1Strings=true`. The comparisons covered all four forms, the Unicode 10 corpus, every BMP code unit, malformed UTF-16, NUL characters, and large and randomized inputs.

### CODE REVIEW

**Review by doistbot — COMMENTED** [review]
This PR swaps the Linux normalization path from a UTF-8 round-trip to a pinned UTF-16 call with a reusable thread-local output buffer, cutting normalization overhead while keeping the public API and results identical across all four forms. No inline issues were flagged during review — reviewers confirmed the pinning, buffer reuse, fallback allocation/free handling, embedded NUL preservation, surrogate handling, and length validation are all correct, and the differential test coverage (malformed UTF-16, buffer boundaries, reuse, concurrency) is thorough. Nice work!

I also included a few optional follow-up notes in the details below.

<details>
<summary>Optional follow-up notes (4)</summary>

- **<a href="#">P3</a> src/linuxMain/kotlin/Normalize.kt:57**: The compatibility requirement noted in the PR description — malformed UTF-16 is only replaced with U+FFFD on libunistring 0.9.7+ — isn't recorded in the repo. On older distros that ship libunistring 0.9.6 or earlier, `u16_normalize` behaves differently than the UTF-8 path for unpaired surrogates. Worth adding a line to README (which mentions the libunistring dependency) and/or a CHANGELOG entry under Unreleased so consumers know the version floor.
- **<a href="#">P3</a> src/linuxTest/kotlin/LinuxNormalizeTest.kt:74**: This sweeps all 65,536 BMP code units, making ~500k native normalization calls. Most of that is identity normalization for valid code points already covered by the NormalizationTest.txt corpus in commonTest; the genuinely new coverage is the unpaired-surrogate range. Narrow the loop to `0xD800..0xDFFF` (plus any specific controls/noncharacters you want) to keep the suite fast without losing the malformed UTF-16 regression signal.
- **<a href="#">P3</a> src/linuxTest/kotlin/LinuxNormalizeTest.kt:156**: This call runs after `results` has already been collected, so it cannot affect the `assertEquals` below and doesn't exercise buffer-reuse interaction. It also duplicates the same assertion in `handlesLargeInputsAndExpandingResults`. Remove it, or move it between the `normalize` calls if the intent was to interleave a larger result before the shorter buffer-reuse cases.
- **<a href="#">P3</a> src/linuxTest/kotlin/LinuxNormalizeTest.kt:184**: This release is unreachable in the scenario the comment describes. Each worker decrements `remaining` as its first action, before anything can fail, so when a job throws no worker is still spinning on the barrier; and if a job were never scheduled, `futures.forEach { assertTrue(it.result) }` would block forever before `finally` ever runs. The `remaining.store(0)` line can be dropped, or the comment rewritten to describe what this actually guards.

</details>

<sub>[Share Feedback](https://github.com/Doist/doistbot-app/issues/new?template=review-feedback.yml&title=Review+feedback+for+Doist%2Fdoistx-normalize&pr_url=https%3A%2F%2Fgithub.com%2FDoist%2Fdoistx-normalize%2Fpull%2F147&datadog_url=https%3A%2F%2Fapp.datadoghq.com%2Flogs%3Fquery%3Dservice%253A%2528doistbot-app-sandbox%2520OR%2520doistbot-app%2529%2520%2540meta.deliveryId%253A8e87e6c0-b28d-11f1-898b-eda7f32eb233%26agg_m%3Dcount%26agg_m_source%3Dbase%26agg_t%3Dcount%26clustering_pattern_field_path%3Dmessage%26cols%3Dhost%252Cservice%26fromUser%3Dtrue%26messageDisplay%3Dinline%26refresh_mode%3Dsliding%26storage%3Dhot%26stream_sort%3Ddesc%26viz%3Dstream%26live%3Dtrue) • [Review Logs](https://app.datadoghq.com/logs?query=service%3A%28doistbot-app-sandbox%20OR%20doistbot-app%29%20%40meta.deliveryId%3A8e87e6c0-b28d-11f1-898b-eda7f32eb233&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=host%2Cservice&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&live=true)</sub>

**Comment by JinneeJ** [comment]
> detekt found more than 20 potential problems in the proposed changes. Check the [Files changed](https://github.com/Doist/doistx-normalize/pull/147/files) tab for more details.

Fixes one warning about the line length. The rest of the warnings usually shouldn't be applicable to tests.


> P3 src/linuxMain/kotlin/Normalize.kt:57: The compatibility requirement noted in the PR description — malformed UTF-16 is only replaced with U+FFFD on libunistring 0.9.7+ — isn't recorded in the repo. On older distros that ship libunistring 0.9.6 or earlier, u16_normalize behaves differently than the UTF-8 path for unpaired surrogates. Worth adding a line to README (which mentions the libunistring dependency) and/or a CHANGELOG entry under Unreleased so consumers know the version floor.

Good call, added the 0.9.7+ requirement to the README setup section and a CHANGELOG entry under Unreleased.


> P3 src/linuxTest/kotlin/LinuxNormalizeTest.kt:74: This sweeps all 65,536 BMP code units, making ~500k native normalization calls. Most of that is identity normalization for valid code points already covered by the NormalizationTest.txt corpus in commonTest; the genuinely new coverage is the unpaired-surrogate range. Narrow the loop to 0xD800..0xDFFF (plus any specific controls/noncharacters you want) to keep the suite fast without losing the malformed UTF-16 regression signal.

I'd rather keep the full sweep. `NormalizationTest.txt` only lists code points with interesting normalization behavior, so it doesn't cover every BMP code unit. This test checks each of them against the previous UTF-8 implementation, including controls and noncharacters, not just the surrogate range. It takes about a second locally, which feels like a fair price for that coverage.

> P3 src/linuxTest/kotlin/LinuxNormalizeTest.kt:156: This call runs after results has already been collected, so it cannot affect the assertEquals below and doesn't exercise buffer-reuse interaction. It also duplicates the same assertion in handlesLargeInputsAndExpandingResults. Remove it, or move it between the normalize calls if the intent was to interleave a larger result before the shorter buffer-reuse cases.

Right, it ran after the results were already collected, so it couldn't affect the assertion. Removed. The first input is already `"é".repeat(4096)`, so the larger result still comes before the shorter reuse cases.

> P3 src/linuxTest/kotlin/LinuxNormalizeTest.kt:184: This release is unreachable in the scenario the comment describes. Each worker decrements remaining as its first action, before anything can fail, so when a job throws no worker is still spinning on the barrier; and if a job were never scheduled, futures.forEach { assertTrue(it.result) } would block forever before finally ever runs. The remaining.store(0) line can be dropped, or the comment rewritten to describe what this actually guards.

I think this one is reachable. The comment is about `worker.execute` failing while jobs are being scheduled, not about a job throwing. If scheduling fails for the third worker, the first two have already decremented `remaining` to 2 and are spinning on the barrier. The exception leaves `mapIndexed` before `futures.forEach`, so `finally` runs next, and without `store(0)` those workers never exit and `requestTermination().result` hangs. It's unlikely in practice, but it keeps the test from hanging instead of failing.

File: src/linuxMain/kotlin/Normalize.kt [hunk 1]
```diff
@@ -1,41 +1,84 @@
...
+                ) ?: return@memScoped this@normalize
+
+                try {
+                    check(length.value <= Int.MAX_VALUE.toULong()) { "Normalized string is too long" }
+                    val resultLength = length.value.toInt()
+                    if (result == output) {
+                        outputBuffer.concatToString(0, resultLength)
+                    } else {
+                        result.readString(resultLength)
+                    }
```
Comment (goncalossilva) [inline src/linuxMain/kotlin/Normalize.kt]: Did you test whether this optimization was really worth it versus going via libunistring every time?

File: src/linuxTest/kotlin/LinuxNormalizeTest.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import kotlin.native.concurrent.TransferMode
+import kotlin.native.concurrent.Worker
+import kotlin.random.Random
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+class LinuxNormalizeTest {
+    @Test
+    fun normalizesHangulSequencesSkippedByCommonCorpusTest() {
```
Comment (goncalossilva) [inline src/linuxTest/kotlin/LinuxNormalizeTest.kt]: Why isn't this in commonTest?

File: src/linuxTest/kotlin/LinuxNormalizeTest.kt [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                assertEquals(composed, input.normalize(Form.NFC))
+                assertEquals(composed, input.normalize(Form.NFKC))
+                assertEquals(decomposed, input.normalize(Form.NFD))
+                assertEquals(decomposed, input.normalize(Form.NFKD))
+            }
+        }
+    }
+
+    @Test
+    fun preservesMalformedUtf16Behavior() {
```
Comment (goncalossilva) [inline src/linuxTest/kotlin/LinuxNormalizeTest.kt]: Same, could be in commonTest?

(same for all below, I won't comment on all)

File: src/linuxTest/kotlin/LinuxNormalizeTest.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    @Test
+    fun handlesEmptyStringsAndEmbeddedNulls() {
+        listOf("", "\u0000", "é\u0000e\u0301\u0000", "😀\u0000").forEach(::assertMatchesUtf8)
+    }
+
+    @Test
+    fun handlesLargeInputsAndExpandingResults() {
+        listOf(4095, 4096, 4097, 16384).forEach { length ->
```
Comment (goncalossilva) [inline src/linuxTest/kotlin/LinuxNormalizeTest.kt]: Good to keep, but this one _almost_ tests implementation details. 😄

**Review by goncalossilva — APPROVED** [review]
LGTM! 👏 

Left a few comments inline, but they boil down to:

- Is the 4K optimization really beneficial in the numbers? If not, I'd drop it, along with the branch and related tests. It'd be simpler!
- I'm not fond of platform-specific tests unless they test platform-specific behaviors. I think most if not all of these tests could be moved to (or dropped in favor of) `commonTest`.

File: src/linuxMain/kotlin/Normalize.kt (hunk 1 above)
Comment (JinneeJ) [reply src/linuxMain/kotlin/Normalize.kt]: TL;DR: removed the buffer :)

So I measured the buffer on its own. Linux ARM64, Kotlin/Native 2.4.0 release, libunistring 1.0-2, five warm-ups and 21 interleaved rounds, with a third variant: pinned UTF-16 without the reusable output buffer, where libunistring allocates the result on every call.

Normalization pair only (NFD → NFC), median ms:

| Dataset | Original UTF-8 | Pinned, no buffer | Pinned + buffer | Buffer gain |
|---|---:|---:|---:|---:|
| Short ASCII | 11.93 | 6.53 | 5.69 | 12.8% |
| Short Cyrillic | 15.71 | 7.34 | 6.42 | 12.5% |
| Short accented Latin | 15.36 | 8.40 | 7.43 | 11.5% |
| Short CJK and emoji | 9.82 | 4.05 | 3.24 | 20.0% |
| Short malformed UTF-16 | 8.94 | 4.91 | 4.13 | 15.8% |
| Mixed text, 512 UTF-16 units | 235.11 | 110.02 | 105.08 | 4.5% |

Full text preparation (NFD → remove combining marks → NFC → lowercase), median ms:

| Dataset | Original UTF-8 | Pinned, no buffer | Pinned + buffer | Buffer gain |
|---|---:|---:|---:|---:|
| Short ASCII | 22.17 | 16.65 | 15.97 | 4.1% |
| Short Cyrillic | 29.72 | 20.95 | 19.99 | 4.6% |
| Short accented Latin | 35.72 | 29.06 | 28.09 | 3.3% |
| Short CJK and emoji | 15.75 | 9.95 | 9.12 | 8.4% |
| Short malformed UTF-16 | 18.08 | 14.05 | 13.34 | 5.1% |
| Mixed text, 512 UTF-16 units | 534.17 | 414.07 | 408.67 | 1.3% |

So the buffer saves 11 to 20% of the normalization pair on short strings, but only 3 to 8% of the whole preparation pipeline, and even less on longer text, where one allocation and copy is small next to the normalization itself. Most of the gain comes from skipping the UTF-8 round-trip, and the pinned path keeps it without the buffer.

For context on what that means outside a microbenchmark, the backend normalizes one or two fields per task while evaluating a filter. On a 5000 task filter, the buffer saves around 0.3 ms out of a 12 to 25 ms evaluation stage.

So saving 0.3 ms in real case doesn't look valid to me to have this buffer, but if you think it might be worth for backend, we can always bring it back.

**Comment by JinneeJ** [comment]
Thanks for the review!

I removed the buffer. Please see the [comment](https://github.com/Doist/doistx-normalize/pull/147#discussion_r4049189403) with details.

Regarding tests, I revisited the testing strategy for this PR and removed the tests that compared the old implementation with the new one. Instead, I added a “Validation” paragraph to the PR description for future investigations. All other tests that could be moved to `commonTest` were moved there, and the remaining ones were removed.
